### PR TITLE
Make the match lowering state machine explicit

### PR DESCRIPTION
**Pull request by Nadrieril**
One of the trickiest parts of match lowering is what happens to the `Candidate`s. They're used for a bunch of things at once, especially for or-patterns. This PR moves the important data of `Candidate` into an enum, to make the various stages of processing explicit.

This is a bit rough, I'm open to feedback on whether this is clear enough. cc @Zalathar in particular.

r? @ghost

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 1]
```diff
@@ -1912,7 +1957,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         // (Only the last candidate can possibly have more match pairs.)
         debug_assert!({
             let mut all_except_last = candidates_to_expand.iter().rev().skip(1);
-            all_except_last.all(|candidate| candidate.match_pairs.is_empty())
+            all_except_last
+                .all(|candidate| matches!(candidate.state, CandidateState::Complete { .. }))
```
Comment (Zalathar) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:1947]: It looks like this is overly strict, because we can also have `CandidateState::Expanded` with `remaining_match_pairs.is_empty()`.

That seems to be the cause of the assertion failure in debug-assertion builds.

**Comment by Zalathar** [comment]
`post_process_expanded_candidates` has ended up being pretty hard to follow.

With the existing code, it's nice that the code that handles `remaining_match_pairs` doesn't care whether it's dealing with an non-trivial or-node, or an or-node that was simplified back into a leaf. Having completely separate code paths for `simplify: true` and `simplify: false` loses some of that simplicity.

**Comment by Nadrieril** [comment]
Yeahh, it's not great, but reusing code as before requires a 4th state I think (see the last commit)

**Comment by Zalathar** [comment]
Hmm yeah, I tried a simplification that I had in mind and it *almost* worked, but not quite.

To me, it feels like the code is not quite ready to be broken up into crisp candidate states, like there's some other change that needs to happen first to make this smoother. But I'm not sure what that change is.

**Comment by Nadrieril** [comment]
I know what you mean.

The other change I have in mind is to choose success/failure blocks _before_ processing candidates, which would mean we can do the "post-processing" of or-patterns before passing them through `match_candidates`. It still doesn't feel enough to clarify the whole thing  though. Also I tried to implement that but the handling of pre_binding_blocks was too tricky.

**Comment by Nadrieril** [comment]
More thought needed then
